src/titiler/xarray/titiler/xarray/io.py

```diff
@@ -0,0 +1,289 @@
+        filesystem = fsspec.filesystem(protocol)  # type: ignore
+        return (
+            filesystem.open(src_path)
+            if xr_engine == "h5netcdf"
+            else filesystem.get_mapper(src_path)
+        )
+
+    else:
+        raise ValueError(f"Unsupported protocol: {protocol}")
+
+
+def xarray_open_dataset(
+    src_path: str,
+    group: Optional[Any] = None,
+    reference: Optional[bool] = False,
+    decode_times: Optional[bool] = True,
+    consolidated: Optional[bool] = True,
+    cache_client: Optional[CacheClient] = None,
+) -> xarray.Dataset:
+    """Open dataset."""
+    # Generate cache key and attempt to fetch the dataset from cache
+    if cache_client:
+        cache_key = f"{src_path}_{group}" if group is not None else src_path
+        data_bytes = cache_client.get(cache_key)
+        if data_bytes:
+            return pickle.loads(data_bytes)
+
+    protocol = parse_protocol(src_path, reference=reference)
+    xr_engine = xarray_engine(src_path)
+    file_handler = get_filesystem(src_path, protocol, xr_engine)
+
+    # Arguments for xarray.open_dataset
+    # Default args
+    xr_open_args: Dict[str, Any] = {
+        "decode_coords": "all",
+        "decode_times": decode_times,
+    }
+
+    # Argument if we're opening a datatree
+    if isinstance(group, int):
+        xr_open_args["group"] = group
+
+    # NetCDF arguments
+    if xr_engine == "h5netcdf":
+        xr_open_args["engine"] = "h5netcdf"
+        xr_open_args["lock"] = False
+    else:
+        # Zarr arguments
+        xr_open_args["engine"] = "zarr"
+        xr_open_args["consolidated"] = consolidated
+
+    # Additional arguments when dealing with a reference file.
+    if reference:
+        xr_open_args["consolidated"] = False
+        xr_open_args["backend_kwargs"] = {"consolidated": False}
+
+    ds = xarray.open_dataset(file_handler, **xr_open_args)
+
+    if cache_client:
+        # Serialize the dataset to bytes using pickle
+        data_bytes = pickle.dumps(ds)
+        cache_client.set(cache_key, data_bytes)
+
+    return ds
+
+
+def _arrange_dims(da: xarray.DataArray) -> xarray.DataArray:
+    """Arrange coordinates and time dimensions.
+
+    An rioxarray.exceptions.InvalidDimensionOrder error is raised if the coordinates are not in the correct order time, y, and x.
+    See: https://github.com/corteva/rioxarray/discussions/674
+
+    We conform to using x and y as the spatial dimension names..
+
+    """
+    if "x" not in da.dims and "y" not in da.dims:
+        latitude_var_name = next(
+            x for x in ["lat", "latitude", "LAT", "LATITUDE", "Lat"] if x in da.dims
+        )
+        longitude_var_name = next(
+            x for x in ["lon", "longitude", "LON", "LONGITUDE", "Lon"] if x in da.dims
+        )
+        da = da.rename({latitude_var_name: "y", longitude_var_name: "x"})
+
+    if "TIME" in da.dims:
+        da = da.rename({"TIME": "time"})
+
+    if _dims := [d for d in da.dims if d not in ["x", "y"]]:
```

Comment:
This is to support dataset with more than 3 dimension (e.g more than time, x, y)